pulsar-client-cpp/lib/MultiTopicsConsumerImpl.cc

```diff
@@ -0,0 +1,621 @@
+                                                       ConsumerConfiguration conf,
+                                                       ConsumerSubResultPromisePtr topicSubResultPromise) {
+    if (result != ResultOk) {
+        LOG_ERROR("Error Checking/Getting Partition Metadata while MultiTopics Subscribing- " << result)
+        topicSubResultPromise->setFailed(result);
+        return;
+    }
+
+    boost::shared_ptr<ConsumerImpl> consumer;
+    ConsumerConfiguration config;
+    ExecutorServicePtr internalListenerExecutor = client_->getPartitionListenerExecutorProvider()->get();
+
+    // all the consumers should have same name.
+    config.setConsumerName(conf_.getConsumerName());
+    config.setConsumerType(conf_.getConsumerType());
+    config.setBrokerConsumerStatsCacheTimeInMs(conf_.getBrokerConsumerStatsCacheTimeInMs());
+    config.setMessageListener(
+        boost::bind(&MultiTopicsConsumerImpl::messageReceived, shared_from_this(), _1, _2));
+    config.setReceiverQueueSize(conf_.getReceiverQueueSize());
+
+    int partitionsNumber = partitionMetadata->getPartitions() >= 1 ? partitionMetadata->getPartitions() : 1;
+
+    Lock lock(mutex_);
+    topicsPartitions_.insert(std::make_pair(topicName->toString(), partitionsNumber));
+    lock.unlock();
+    allTopicPartitionsNumber_->fetch_add(partitionsNumber);
+
+    boost::shared_ptr<std::atomic<int>> partitionsNeedCreate =
+        boost::make_shared<std::atomic<int>>(partitionsNumber);
+
+    for (int i = 0; i < partitionsNumber; i++) {
+        std::string topicPartitionName = topicName->getTopicPartitionName(i);
+        consumer = boost::make_shared<ConsumerImpl>(client_, topicPartitionName, subscriptionName_, config,
+                                                    internalListenerExecutor, Partitioned);
+        consumer->getConsumerCreatedFuture().addListener(
+            boost::bind(&MultiTopicsConsumerImpl::handleSingleConsumerCreated, shared_from_this(), _1, _2,
+                        partitionsNeedCreate, topicSubResultPromise));
+        consumer->setPartitionIndex(i);
+        consumers_.insert(std::make_pair(topicPartitionName, consumer));
+        LOG_DEBUG("Create Consumer for - " << topicPartitionName << " - " << consumerStr_);
+        consumer->start();
+    }
+}
+
+void MultiTopicsConsumerImpl::handleSingleConsumerCreated(
+    Result result, ConsumerImplBaseWeakPtr consumerImplBaseWeakPtr,
+    boost::shared_ptr<std::atomic<int>> partitionsNeedCreate,
+    ConsumerSubResultPromisePtr topicSubResultPromise) {
+    if (state_ == Failed) {
+        // one of the consumer creation failed, and we are cleaning up
+        topicSubResultPromise->setFailed(ResultAlreadyClosed);
+        LOG_ERROR("Unable to create Consumer - state == Failed, result: " << result);
+        return;
+    }
+
+    int previous = partitionsNeedCreate->fetch_sub(1);
+    assert(previous > 0);
+
+    if (result != ResultOk) {
+        state_ = Failed;
```

Comment:
state shouldn't be set here. It will be set when the failure on the promise propagates up.